### PR TITLE
Handle parsing envID & VPid from process status file

### LINKED ISSUE #55: CTID doesn't show openvz container id
Description of problem:
CTID doesn't show openvz container id

Version-Release number of selected component (if applicable):
htop-2.2.0-3.el7.x86_64

How reproducible:
always

Steps to Reproduce:
1.  running on vzkernel-3.10.0-862.20.2.vz7.73.29.x86_64
2.  start container(s)
3.  launch htop (on HN)
4.  show CTID column

Actual results:
CTID 0 for all processes

Expected results:
correct container IDs

for further details see comments in:  https://bugzilla.redhat.com/show_bug.cgi?id=1694265

### LINKED ISSUE #192: Fishy fscanf statement for OpenVZ data retrieval
In the below code snippet the format string is kinda bad style and should be replaced by something more sensible.

https://github.com/htop-dev/htop/blob/9ee72568dc36e2c83add9c602387509909f3d292/linux/LinuxProcessList.c#L499-L507

Furthermore the size given (32) for most arguments is far outside of sensible values for `%d` (sensible values would be up to 20 when considering `int` to map to `int64_t`).

Is there any documentation to point to for the format of the file we try to read there from OpenVZ? Couldn't find anything relevant (also no OpenVZ available for testing).

### PR DESCRIPTION
**Pull request by BenBE**
Fixes #55
Fixes #192

### CODE REVIEW

File: linux/LinuxProcess.h [hunk 1]
```diff
@@ -120,8 +120,8 @@ typedef struct LinuxProcess_ {
    double io_rate_write_bps;
    #endif
    #ifdef HAVE_OPENVZ
-   unsigned int ctid;
-   unsigned int vpid;
+   char* ctid;
```
Comment (cgzones) [inline linux/LinuxProcess.h:123]: This new dynamic allocation needs to be cleaned up in
https://github.com/htop-dev/htop/blob/241e4b3dbf8ebee4a12a337dec725b3547c242e9/linux/LinuxProcess.c#L145-L153

File: linux/LinuxProcess.c [hunk 2]
```diff
@@ -251,7 +251,7 @@ void LinuxProcess_writeField(Process* this, RichString* str, ProcessField field)
    }
    #endif
    #ifdef HAVE_OPENVZ
-   case CTID: xSnprintf(buffer, n, "%7u ", lp->ctid); break;
+   case CTID: xSnprintf(buffer, n, " %s ", lp->ctid ? lp->ctid : ""); break;
```
Comment (cgzones) [inline linux/LinuxProcess.c:254]: Probably use a minimal width here

**Review by cgzones — CHANGES_REQUESTED** [review]
Code review only, not actually tested.

File: linux/LinuxProcess.c [hunk 3]
```diff
@@ -77,7 +77,7 @@ ProcessFieldData Process_fields[] = {
    [NLWP] = { .name = "NLWP", .title = "NLWP ", .description = "Number of threads in the process", .flags = 0, },
    [TGID] = { .name = "TGID", .title = "   TGID ", .description = "Thread group ID (i.e. process ID)", .flags = 0, },
 #ifdef HAVE_OPENVZ
-   [CTID] = { .name = "CTID", .title = "   CTID ", .description = "OpenVZ container ID (a.k.a. virtual environment ID)", .flags = PROCESS_FLAG_LINUX_OPENVZ, },
+   [CTID] = { .name = "CTID", .title = " CTID ", .description = "OpenVZ container ID (a.k.a. virtual environment ID)", .flags = PROCESS_FLAG_LINUX_OPENVZ, },
```
Comment (cgzones) [inline linux/LinuxProcess.c:80]: Please use .title = " CTID&nbsp;&nbsp;&nbsp;&nbsp;" (4 trailing spaces) to ensure correct alignment

File: linux/LinuxProcess.c [hunk 4]
```diff
@@ -253,7 +256,7 @@ void LinuxProcess_writeField(Process* this, RichString* str, ProcessField field)
    }
    #endif
    #ifdef HAVE_OPENVZ
-   case CTID: xSnprintf(buffer, n, "%7u ", lp->ctid); break;
+   case CTID: xSnprintf(buffer, n, " %-8s ", lp->ctid ? lp->ctid : ""); break;
```
Comment (cgzones) [inline linux/LinuxProcess.c:259]: Drop the trailing blank (`"%-8s "`)

File: linux/LinuxProcessList.c [hunk 5]
```diff
@@ -487,25 +487,94 @@ static bool LinuxProcessList_readSmapsFile(LinuxProcess* process, const char* di
...
+         continue;
+      }
+
+      do {
+         name_value_sep++;
+      } while(*name_value_sep && *name_value_sep <= 32);
+
+      char* value_end = name_value_sep;
+
+      while(*value_end && *value_end > 32) {
```
Comment (cgzones) [inline linux/LinuxProcessList.c:540]: `*value_end` (= `*value_end != 0`) seems redundant when checking for `*value_end > 32`

File: linux/LinuxProcessList.c [hunk 6]
```diff
@@ -487,25 +487,94 @@ static bool LinuxProcessList_readSmapsFile(LinuxProcess* process, const char* di
...
+         field = 1;
+      } else if(0 == strncasecmp(linebuf, "VPid", name_value_sep - linebuf)) {
+         field = 2;
+      } else {
+         continue;
+      }
+
+      do {
+         name_value_sep++;
+      } while(*name_value_sep && *name_value_sep <= 32);
```
Comment (cgzones) [inline linux/LinuxProcessList.c:536]: Maybe use `*name_value_sep != '\0'` to make it more clear that we stop on nul byte

File: linux/LinuxProcessList.c [hunk 7]
```diff
@@ -487,25 +487,94 @@ static bool LinuxProcessList_readSmapsFile(LinuxProcess* process, const char* di
...
+
+      while(*value_end && *value_end > 32) {
+         value_end++;
+      }
+
+      if(name_value_sep == value_end) {
+         continue;
+      }
+
+      *value_end = 0;
```
Comment (cgzones) [inline linux/LinuxProcessList.c:548]: Maybe use more descriptive nul byte `'\0'`

File: linux/LinuxProcessList.c [hunk 8]
```diff
@@ -487,25 +487,94 @@ static bool LinuxProcessList_readSmapsFile(LinuxProcess* process, const char* di
...
+            free(process->ctid);
+            process->ctid = xStrdup(name_value_sep);
+         }
+         break;
+      case 2:
+         foundVPid = true;
+         process->vpid = strtoul(name_value_sep, NULL, 0);
+         break;
+      default:
+         //Sanity Check: Should never reach here, or the implementation is missing something!
```
Comment (cgzones) [inline linux/LinuxProcessList.c:563]: I prefer either dropping this block or maybe add an `assert(0);`

File: linux/LinuxProcessList.c (hunk 5 above)
Comment (BenBE) [reply linux/LinuxProcessList.c:540]: I included it to make the end-of-string check more obvious.

File: linux/LinuxProcessList.c (hunk 8 above)
Comment (BenBE) [reply linux/LinuxProcessList.c:563]: Will add an `assert(0);` for it.

**Comment by BenBE** [comment]
Does column resizing work on-the-fly (most strings are short ints, but could also be long UUIDs)?

File: linux/LinuxProcess.c (hunk 4 above)
Comment (BenBE) [reply linux/LinuxProcess.c]: You mean the leading?

**Review by cgzones — APPROVED** [review]

File: linux/LinuxProcessList.c (hunk 5 above)
Comment (cgzones) [reply linux/LinuxProcessList.c]: Thanks

File: linux/LinuxProcess.c (hunk 4 above)
Comment (cgzones) [reply linux/LinuxProcess.c]: Yes, sorry for the confusion

**Review by cgzones — APPROVED** [review]

**Comment by cgzones** [comment]
Thanks, applied.

**Pull request merged**